### PR TITLE
Mark IeeeFlags and FloatingPointControl as @trusted nothrow @nogc

### PR DESCRIPTION
**Pull request by ibuclaw**
Have added `pure` onto the flag getters, and removed `pure` from the flag setters (for now...).

Granted that `std.math.exp` is pure _and still_ raises an FP exception gives us at least an argument for adding pure to these structures also (or removing the `pure`  attribute from this and others functions, breaking everyone's code).

As well, a new private `FloatingPointControl.raiseException` method should be added too.  As this hack can now be defeated by the optimizer, however this wasn't the case in the past.
https://github.com/dlang/phobos/blob/fd9fdeb25d3f497bf68ba5ac29667adb3c912ddb/std/math.d#L1833-L1850

Edit: Although now I'm swayed in the direction that the above should just be removed and unittests updated.

### CODE REVIEW

**Comment by JackStouffer** [comment]
Pinging qualified reviewers @kinke @klickverbot

File: std/math.d [hunk 1]
```diff
@@ -5225,7 +5233,7 @@ public:
     }
 
     /// Returns: the exceptions which are currently enabled (unmasked)
-    @property static ExceptionMask enabledExceptions() @nogc
+    @property static ExceptionMask enabledExceptions()
```
Comment (kinke) [inline std/math.d]: Should be pure as well.

File: std/math.d [hunk 2]
```diff
@@ -5054,14 +5062,14 @@ struct FloatingPointControl
...
-    @property void rounding(RoundingMode newMode) @nogc
+    @property void rounding(RoundingMode newMode)
     {
         initialize();
         setControlState((getControlState() & (-1 - roundingMask)) | (newMode & roundingMask));
     }
 
     /// Returns: the currently active rounding mode
-    @property static RoundingMode rounding() @nogc
+    @property static RoundingMode rounding()
```
Comment (kinke) [inline std/math.d]: pure

File: std/math.d [hunk 3]
```diff
@@ -4950,10 +4953,13 @@ else version(ARM)
...
 /// Set all of the floating-point status flags to false.
-void resetIeeeFlags() @nogc { IeeeFlags.resetIeeeFlags(); }
+void resetIeeeFlags() @trusted nothrow @nogc
+{
+    IeeeFlags.resetIeeeFlags();
+}
 
 /// Returns: snapshot of the current state of the floating-point status flags
-@property IeeeFlags ieeeFlags()
+@property IeeeFlags ieeeFlags() @trusted nothrow @nogc
```
Comment (kinke) [inline std/math.d]: pure; possibly `@safe` (same for `resetIeeeFlags()` above)

**Comment by kinke** [comment]
LGTM otherwise, thanks.

File: std/math.d [hunk 4]
```diff
@@ -4789,11 +4791,12 @@ private:
         else
             assert(0, "Not yet supported");
     }
-    static void resetIeeeFlags() @nogc
+
+    static void resetIeeeFlags()
```
Comment (n8sh) [inline std/math.d]: Should be `pure` per the specification:
https://dlang.org/spec/function.html#pure-functions
>8. As a concession to practicality, a pure function can also:
> * read **and write** the floating point exception flags
> ...

File: std/math.d [hunk 5]
```diff
@@ -5280,18 +5288,18 @@ private:
     }
 
     // Clear all pending exceptions
-    static void clearExceptions() @nogc
+    static void clearExceptions()
```
Comment (n8sh) [inline std/math.d]: `pure`

**Comment by ibuclaw** [comment]
Rebased and added pure onto more getters.  As you can see from @n8sh's comments, not the setters just yet.

**Pull request merged**